### ISSUE TITLE
token.left_edge/right_edge gets out of sync when manually modifying doc heads

### ISSUE THREAD

**Issue by jessecoleman** [explosion/spaCy#4497]
I'm doing some post processing on the output from the `parse` pipe. As part of this process, I generate multiple document roots by setting `doc[i].head = doc[i]`. I would expect this to recreate the dependency tree and recompute all the associated attributes. It works as expected when calling `.subtree`, `.ancestors`, `.lefts`, and `.rights`, but `.left_edge` and `.right_edge` use the old tree structure.

## How to reproduce the behaviour
```
import spacy
en = spacy.load('en_core_web_sm')
doc = en('apple pie')
assert doc[1].left_edge == doc[0]
doc[0].head = doc[0]
assert list(doc[1].subtree) == [doc[1]] # fine
assert list(doc[1].lefts) == [] # fine
assert list(doc[0].ancestors) == [] # fine
assert doc[1].left_edge == doc[1] # AssertionError
```

## Your Environment
* Operating System: Ubuntu 16.04
* Python Version Used: 3.6.8
* spaCy Version Used: 2.2.1

**Comment by jessecoleman** [comment]
Maybe related? https://github.com/explosion/spaCy/issues/251

**Comment by honnibal** [comment]
Interesting, this must be quite an old bug!

I would've expected `Token.head.__set__` to call into the `set_children_from_heads` function that's in `spacy.tokens.doc`. Instead it seems to have its own code in there.

I haven't looked in detail, but my hunch is that the code in the `Token.head.__set__` assumes the parse tree is projective, and so does the wrong thing.

**Comment by adrianeboyd** [comment]
I'm not sure it's related to projectivity. I think it's not dealing with adding a new root / separate tree correctly. The sent starts are also not updated correctly.

I can hackily fix this particular problem by detecting whether this is setting a new root and using `set_children_from_heads()` instead for that case, but what I don't understand yet is why replacing most of the code here with `set_children_from_heads()` doesn't work for the very non-projective test case in the test suite.

**Comment by adrianeboyd** [comment]
I did figure this out, I think, in #6046, which replaces this code with `set_children_from_heads`.

I hope this is fixed in v3.0, please open a new issue if you run into problems!
